Make circle (shift-oval) tool work for any direction dragged

towerofnix opened the pull request:
### Resolves

Fixes #710.

### Proposed Changes

Changed math used when shift key is pressed while creating an oval so that the circle can be drawn up-right and bottom-left of where the drag started.

### Reason for Changes

Bringing over features of Scratch 2.0. Note that dragging still isn't quite the same as in 2.0 (3.0 is based exclusively on the `x` offset instead of both); but with this PR, at least you *can* drag to the four relative quadrants rather than just two.

### Test Coverage

Tested manually. This math will need to be moved into a utility function, so that the bitmap oval tool works too. We might want to tweak the `diameter` variable to work however Scratch 2.0 does (try comparing the two versions yourself and you'll *feel* the difference); likewise, we might want to make the rectangle tool use this same math, so that rectangles scale the same way as ovals.

towerofnix committed:
Make circle tool work for any relative quadrant

fsih commented on src/helper/tools/oval-tool.js:
```diff
@@ -89,7 +89,21 @@ class OvalTool extends paper.Tool {
+            let offsetX = event.point.x - event.downPoint.x;
+            let offsetY = event.point.y - event.downPoint.y;
+
+            // If the offset variables are zero, the circle ends up having zero width or height, which is bad.
```
This step isn't necessary because if you look at line 128, if you draw an oval that's too small it will just get removed.

towerofnix replied:
I think the problem was that if you drag a circle out to, for example, X = 100 and Y = 0 (relative to where you started dragging), you would *expect* to see a 100-unit circle, but there is none (because offsetX/Y = 0). The actual case where this is jarring is when you go from Y = 1 to Y = 0, because then the 100-unit (from X=100, Y=1) circle disappears.

fsih reviewed (commented):
Thanks! This is definitely better than what we have now.

I agree that it would be nice to share this code with the bit oval tool. Code shared between tools is somewhat arbitrarily split up into helper js files right now. Maybe it would make sense in helper/math? You could also make a new helper file.

I also agree that it would be better if the behavior when holding shift worked more like it does in Scratch 2. I think it's making the diameter be the larger of offset x and offset y.

If you want to move on from this PR without adding those changes though that's fine, I still think we should merge it since it's incrementally better.

towerofnix committed:
Separate "square dimensions" math into shared file

towerofnix commented:
Alright, I split the math code into `helper/math.js` per your suggestion. This definitely helps clean up the code and makes it easier to tweak the math (e.g. to make it work like 2.0) in all places it's used.

In a similar vein of thought, I noticed I had to make a few identical changes to both `oval-tool.js` and `rect-tool.js` (which I implemented to make sure that it was useful across both tools). Maybe this is a little bit overkill, but I don't think so - Would it be useful to separate *all* the code which determines the size and position of a drawn shape into a shared math function? Ala `getShapeDimensions(startPos, eventPoint, modifiers)`; it would take into account alt, shift, and neither being pressed so that the other files don't have to have all that duplicate code themselves.

towerofnix committed:
Make square scaling work like Scratch 2.0

fsih commented on src/helper/math.js:
```diff
@@ -131,13 +131,42 @@ const scaleWithStrokes = function (root, factor, pivot) {
     root.scale(factor, pivot);
 };
 
+// Get the size and position of a square, as in if the user were holding the shift key down while drawing the shape,
```
Could you switch this to a JS Doc, with
```
/**
 * @param
 * @returns
 */
```
? I know the rest of the code in this file doesn't have proper docs, but we can start here!

towerofnix replied:
Done (86c3bb1a4f4a1597a4fd04097dfbaa95b0758c89).

fsih reviewed (changes requested):
The new algorithm feels really good!

Do you mind switching  the bitmap oval and rect tools over to this algorithm too?

fsih commented:
> In a similar vein of thought, I noticed I had to make a few identical changes to both oval-tool.js and rect-tool.js

Yeah, those two tools are almost identical. I have been meaning to make them share a base class that they both extend (say ShapeTool). I'll probably refactor them if we add more shapes.

towerofnix committed:
Implement square scaling in bitmap tools

towerofnix committed:
JSDoc-ify description of getSquareDimensions

towerofnix commented:
> Do you mind switching the bitmap oval and rect tools over to this algorithm too?

Sure! Done, see 7e8bb8dfaf0b421060fe601cc778b512e609ce3d. Also JSDoc-ified that comment (86c3bb1a4f4a1597a4fd04097dfbaa95b0758c89).

fsih commented:
This looks great, thanks so much @towerofnix!
